stdlib/public/runtime/SymbolInfo.cpp

```diff
@@ -59,42 +60,119 @@ const char *SymbolInfo::getSymbolName() const {
 
 const void *SymbolInfo::getSymbolAddress() const {
 #if defined(_WIN32) && !defined(__CYGWIN__)
-  return reinterpret_cast<const void *>(_package.si.Address);
+  return _symbolAddress;
 #elif SWIFT_STDLIB_HAS_DLADDR
   return _info.dli_saddr;
 #else
   return nullptr;
 #endif
 }
 
-std::optional<SymbolInfo> SymbolInfo::lookup(const void *address) {
-  std::optional<SymbolInfo> result;
+#if defined(_WIN32) && !defined(__CYGWIN__)
+struct Win32ModuleInfo {
+  const char *name;
+  const void *base;
+
+  Win32ModuleInfo(const char *n, const void *b) : name(n), base(b) {}
+};
+
+// Get the filename and base of the module that contains the specified
+// address.  N.B. This returns a `malloc`-ed copy of the filename in the
+// Win32ModuleInfo struct; you are responsible for freeing that.
+static Win32ModuleInfo moduleInfoFromAddress(const void *address) {
+  HMODULE hModule;
+  if (GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS
+                        | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
+                         (LPCTSTR)address,
+                         &hModule)) {
+    MODULEINFO mi;
+
+    if (!GetModuleInformation(GetCurrentProcess(), hModule,
+                              &mi, sizeof(mi))) {
+      ZeroMemory(&mi, sizeof(mi));
+    }
+
+    WCHAR wszBuffer[256];
+    LPWSTR pwszFileName = wszBuffer;
```

Comment:
Can we use `std::unique_ptr<WSTR>`? That would simplify the exit path.

Reply:
The default destructor uses global operator delete I think, which is something you'd really want to avoid.

Reply:
How about `std:array`?

Reply:
`std::unique_ptr` takes an optional custom deleter, which can use `free`. It is just a tad more verbose to write:

```c++
std::unique_ptr<WSTR, decltype(std::free)> buffer{nullptr, std::free};
buffer.reset(std::realloc(buffer.release(), size << 1));
```